syntax = "proto3";

package envoy.extensions.filters.http.set_metadata.v3;

import "google/protobuf/any.proto";
import "google/protobuf/struct.proto";

import "envoy/annotations/deprecation.proto";
import "udpa/annotations/status.proto";
import "validate/validate.proto";

option java_package = "io.envoyproxy.envoy.extensions.filters.http.set_metadata.v3";
option java_outer_classname = "SetMetadataProto";
option java_multiple_files = true;
option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/set_metadata/v3;set_metadatav3";
option (udpa.annotations.file_status).package_version_status = ACTIVE;

// [#protodoc-title: Set-Metadata Filter]
//
// This filters adds or updates dynamic metadata with static data.
//
// [#extension: envoy.filters.http.set_metadata]

message Config {
  message UntypedMetadata {
    // The metadata namespace.
    string metadata_namespace = 1 [(validate.rules).string = {min_len: 1}];

    // The value to place at the namespace. If ``allow_overwrite``, this will
    // overwrite or merge with any existing values in that namespace.
    google.protobuf.Struct value = 2;

    // Allow the filter to overwrite or merge with an existing value in the namespace.
    bool allow_overwrite = 3;
  }

  message TypedMetadata {
    // The metadata namespace.
    string metadata_namespace = 1 [(validate.rules).string = {min_len: 1}];

    // The value to place at the namespace. If ``allow_overwrite``, this will
    // overwrite any existing values in that namespace. See
    // :ref:`the filter documentation <config_http_filters_set_metadata>` for
    // more information on how this value is merged with potentially existing
    // ones if ``allow_overwrite`` is configured.
    google.protobuf.Any value = 2;

    // Allow the filter to overwrite an existing value in the namespace.
    bool allow_overwrite = 3;
  }

  // The metadata namespace.
  // metadata_namespace is deprecated; use untyped_metadata or typed_metadata
  string metadata_namespace = 1
      [deprecated = true, (envoy.annotations.deprecated_at_minor_version) = "3.0"];

  // The untyped value to update the dynamic metadata namespace with. See
  // :ref:`the filter documentation <config_http_filters_set_metadata>` for
  // more information on how this value is merged with potentially existing
  // ones.
  // value is deprecated; to configure a filter to set untyped dynamic metadata,
  // use the untyped_metadata field. To configure a filter to set typed dynamic
  // metadata, use the typed_metadata field.
  google.protobuf.Struct value = 2
      [deprecated = true, (envoy.annotations.deprecated_at_minor_version) = "3.0"];

  // Defines changes to be made to untyped dynamic metadata.
  repeated UntypedMetadata untyped_metadata = 3;

  // Defines changes to be made to typed dynamic metadata.
  repeated TypedMetadata typed_metadata = 4;
}
